README

ncm-ncd: front end for executing NCM components.

Author: German Cancio <German.Cancio@cern.ch>

quattro management toolsuite:
http://cern.ch/quattro



ChangeLog
2011-10-21  Loic Brarda

	* Release: 1.5.0
	- Add option to set the syslog facility (defaults to local1)

2011-06-30 Ulrich Schwickerath
        * Release: 1.4.1-2
        - decouple the CERN-CC specific packages from the exact spma version
        - add support for configuration of randomization in CERN-CC configuration files

2011-05-09  Nick Williams

	* Release: 1.4.1
	- add in some taint checks given that ncm-ncd now is in taint-warn mode

2011-05-09  Nick Williams

	* Release: 1.4.0
	- bugfix: when running in NoAction mode, don't modify any state files

2011-03-30  "Belmiro Moreira" "Ulrich Schwickerath"

        * Release: 1.3.4
        - Skip a list of components (comma separated)
	- add -t option to ncm-ncd to issue security warnings

2011-03-23  Michel Jouvin

	* Release: 1.3.3
	- Fix escape() method (not declared as a class method)

2010-09-01  Nick Williams

	* Release: 1.3.2
	- second attempt at checkin

2010-09-01  Nick Williams

	* Release: 1.3.1
	- Added a prefix() method to make subclassing components easier

2010-06-10  "Nick Williams"

	* Release: 1.3.0
	- enable state directory to track which components are running/failed

2008-10-21  "Luis Fernando"

	* Release: 1.2.26
	- NCD::ComponentProxyList should not call setup_reporter().
	
	This module is not a CAF::ReporterMany but a CAF::Reporter. That call was overwriting
	the settings from the command line, leading to --debug and --verbose not to work.

2008-09-26  "Luis Fernando"

	* Release: 1.2.25
	- Making NCD/* inherit from CAF::ReporterMany, which keeps them thread-safe.
	Needs perl-CAF 1.6.5 or higher.

2008-09-21  "Ulrich Schwickerath"

	* Release: 1.2.24
	- replaced obsolete regis01 server name in CERN-CC parts

2008-07-28  "Luis Fernando"

	* Release: 1.2.23
	- Fixed bug #39246, --multilog works again with perl-CAF 1.4.15 and higher.

2008-07-07  "Luis Fernando"

	* Release: 1.2.22
	- Sanitized the component's name on NCD::ComponentProxy so that it runs without warnings with -t.

2008-06-16  Marco Emilio Poleggi,<Marco.Poleggi@cnaf.infn.it>

	* Release: 1.2.21
	- NCM::Check extended with 'noaction' option for overriding the global flag

2007-12-21  German Cancio Melia

	* Release: 1.2.20
	- Fix for bug #20040: add 'NOACTION_SUPPORTED' to NCM component

2007-12-21  German Cancio Melia

	* Release: 1.2.19
	- Added --skip option (Savannah #20619)

2007-12-06  Marco Emilio Poleggi,<Marco.Poleggi@cern.ch>

	* Release: 1.2.18
	- Previous change rolled back, since the issue with CAF::Reporter 
	is better solved by the last patch in CAF::Object (rel. 1.4.15)

2007-12-05  Marco Emilio Poleggi,<Marco.Poleggi@cern.ch>

	* Release: 1.2.17
	- ComponentProxy: fix for new Reporter class which needs to be properly 
	initialized.

2007-02-28  German Cancio Melia

	* Release: 1.2.16
	- REALLY Log ccm-fetch information into log file when running ncm_wrapper

2007-02-28  German Cancio Melia

	* Release: 1.2.15
	- Log ccm-fetch information into log file when running ncm_wrapper

2006-11-28  German Cancio Melia

	* Release: 1.2.14
	- Fix for bug #21956: "ncm-ncd --list" should not try to acquire a lock

2006-11-17  German Cancio Melia

	* Release: 1.2.13
	- Make 'backup' option truly optional in NCM::Check::lines (thanks to L. Cons)

2006-11-06  German Cancio Melia

        * Release: 1.2.12
        - Do not ignore CHLD signal during component execution

2006-09-22  German Cancio Melia

        * Release: 1.2.11
        - Ensure signals to be ignored are ignored, including SIGHUP

2006-08-02  German Cancio Melia

        * Release: 1.2.10
        - Just to be *completely* paranoiac: disallow *any* combination of simultaneous invocations, whatever the flags are, by locking also when NCM is invoked in --noaction mode!

2006-08-02  German Cancio Melia

        * Release: 1.2.9
        - Do not allow NCM to run in --noaction mode if already running in 'real' mode
        - Fix Savannah bug# 18678 (wrong cyclic dependencies detected when there are none)

2006-08-01  German Cancio Melia

        * Release: 1.2.8
        - Added dependency on ncm-template; fix Savannah bug #18631 (restore template delimiters if changed by a component)

2006-05-26  German Cancio Melia

        * Release: 1.2.7
        - (add tag to release)

2006-05-25  Stephen Childs

        * Release: 1.2.7
        - Fix Savannah #16822 (print names of faulty components at end of run)
2005-10-08  Charles Loomis

        * Release: 1.2.6
        - fix bad substitutions

2005-10-08  Charles Loomis

        * Release: 1.2.5
        - remove spurious debugging statements

2005-10-08  Charles Loomis

        * Release: 1.2.4
        - allow components to be defined in XML profile

2005-05-20  Juan Antonio LOPEZ PEREZ

        * Release: 1.2.3
        - Substitution of edg-ccm-fetch by ccm-fetch. The script changed the name in the latest versions

2005-05-11  German Cancio Melia

        * Release: 1.2.2
        - (CERN-CC only) splitted up notd entries for ncm into ncm and ncm_unconf

2005-05-07  German Cancio Melia

        * Release: 1.2.1
        - Fix Savannah #3180 (added lockfile for application and corresponding options)

2005-04-18  German Cancio Melia

        * Release: 1.1.22
        - Fix Savannah bugs #7981

2005-04-04  German Cancio Melia

        * Release: 1.1.21
        - Updated HLConfig.pm (Jan van Eldik)

2005-03-31  German Cancio Melia

        * Release: 1.1.20
        - Added escape() function to NCM::Component

2005-01-21  German CANCIO MELIA

        * Release: 1.1.19
        - Fix Savannah #6281 - more verbose logging of broken dependencies

2004-11-18  German CANCIO MELIA

        * Release: 1.1.18
        - Added new variables for OS detection within the components: $SYSNAME and $SYSVERS

2004-10-28  Rafael Garcia Leiva

        * Release: 1.1.17
        - updated dependencies

2004-09-07  German Cancio Melia

        * Release: 1.1.16
        - Added waittofin = yes to CERN CC notd wrapper

2004-09-04  German CANCIO MELIA


src/main/perl/ComponentProxy.pm
# ${license-info}
# ${developer-info
# ${author-info}
# ${build-info}


package NCD::ComponentProxy;

use strict;
use vars qw(@ISA $this_app);
use CAF::ReporterMany;
use LC::Exception qw (SUCCESS throw_error);

use CAF::Object;
use EDG::WP4::CCM::CacheManager;
use EDG::WP4::CCM::Path;
use CAF::Log;
use LC::Check;
use NCM::Template;

use File::Path;

*this_app = \$main::this_app;


@ISA=qw(CAF::Object CAF::ReporterMany);


my $_COMP_PREFIX='/software/components';

my $ec=LC::Exception::Context->new->will_store_all;

# default template delimiters (will be restored before running a component)
my @_TEMPLATE_DELIMITERS=NCM::Template->GetDelimiters();

=pod

=head1 NAME

NCD::ComponentProxy - component proxy class

=head1 SYNOPSIS


=head1 INHERITANCE

  CAF::Object, CAF::Reporter

=head1 DESCRIPTION

Provides management functions for accessing and executing NCM
components.

=over

=back

=head1 AUTHOR

German Cancio <German.Cancio@cern.ch>

=head1 VERSION

$Id: ComponentProxy.pm.cin,v 1.34 2008/09/26 15:59:34 munoz Exp $

=cut


#------------------------------------------------------------
#                      Public Methods/Functions
#------------------------------------------------------------

=pod

=back

=head2 Public methods

=over 4

=item executeConfigure(): ref(hash)

executeConfigure loads and executes the component with
'configure'. the number of produced errors and warnings is returned in
a hash ('ERRORS','WARNINGS'). If the component cannot be executed,
undef is returned.

=cut

sub executeConfigure {
  my $self=shift;

  return $self->_execute('Configure');
}


=pod

=item executeUnconfigure(): ref(hash)

executeUnconfigure loads and executes the component with
'unconfigure'. the number of produced errors and warnings is returned
in a hash ('ERRORS','WARNINGS'). If the component cannot be executed,
undef is returned.


=cut

sub executeUnconfigure {
  my $self=shift;

  return $self->_execute('Unconfigure');

}


=pod

=item name(): string

returns the name of the component

=cut

sub name {
  my $self=shift;
  return $self->{'NAME'};
}


=pod

=item getPreDependencies(): ref(@array)

returns an array to the names of predependent components. The array is
empty if no predependencies are found.

=cut

sub getPreDependencies {
  my $self=shift;

  return $self->{'PRE_DEPS'};
}

=pod

=item getPostDependencies(): ref(@array)

returns an array to the names of postdependent components. The array is
empty if no postdependencies are found.

=cut

sub getPostDependencies {
  my $self=shift;

  return $self->{'POST_DEPS'};
}


=pod

=item hasFile(): boolean

returns 1 if the components perl module is installed, 0 otherwise.

=cut

sub hasFile {
  my $self=shift;

  return -r '/usr/lib/perl/NCM/Component/'.$self->name().'.pm' ? 1:0 ;
}

=pod
=item writeComponent(): boolean

returns 1 if the component has the code defined in the configuration
and has been written to disk, 0 otherwise.  This will erase old
component definitions if the code is no longer defined in the
XML configuration.

=cut

sub writeComponent {

  my $self=shift;

  # Pull out the component name and the configuration.
  my $cname = $self->{'NAME'};
  my $config = $self->{'CONFIG'};

  # Ensure that both the name and configuration are defined.
  unless (defined($cname)) {
      $self->error("internal error: component name not defined");
  }
  unless (defined($config)) {
      $self->error("internal error: component configuration not defined");
  }

  # Base name for component configuration.
  my $base = '/software/components/'.$cname;

  # Determine if the script exists.  If not, then ensure that any
  # files created by previous runs are removed.  This is to avoid
  # interference between old scripts and newer ones installed via
  # a package.  This is needed because there is no hook for
  # cleaning up a script if no 'Unconfigure' method is defined.
  unless ($config->elementExists($base.'/code/script')) {

      # If the script exists, remove it.
      my $fname = "/var/ncm/lib/perl/NCM/Component/$cname.pm";
      if (-e $fname) {
	  unlink $fname;
	  $self->error("error unlinking $fname") if ($?);
      }

      # Remove data directory for this template.
      my $dname = "/var/ncm/config/$cname";
      rmtree($dname,0,1) if (-e $dname);

      return 0;
  }

  # Ensure that the directory for the components exists.
  my $sdir = '/var/ncm/lib/perl/NCM/Component';
  unless (-d $sdir) {
      mkpath($sdir, 0, 0755);
      unless (-d $sdir) {
	  $self->error("cannot create directory: $sdir");
	  return 0;
      }
  }

  # Ensure that the directory for the component data exists.
  my $ddir = '/var/ncm/config/'.$cname;
  unless (-d $ddir) {
      mkpath($ddir, 0, 0755);
      unless (-d $ddir) {
	  $self->error("cannot create directory: $ddir");
	  return 0;
      }
  }

  # Now write the script to the file.
  my $script = $config->getValue($base.'/code/script');
  my $fname = "$sdir/$cname.pm";
  open SCRIPT, '>', "$fname";
  print SCRIPT $script;
  close SCRIPT;

  # Check if there was an error while writing the script.
  if ($?) {
      $self->error("error writing script $fname: $!");
      return 0;
  }

  # Write out data files if specified.
  if ($config->elementExists($base.'/code/data')) {
      my $dhash = $config->getElement($base.'/code/data');
      while ($dhash->hasNextElement()) {
	  my $entry = $dhash->getNextElement();
	  my $fname = $entry->getName();
	  my $contents = $config->getValue($base.'/code/data/'.$fname);

	  # Now write the script to the file.
	  open DATA, '>', "$ddir/$fname";
	  print DATA $contents;
	  close DATA;

	  # Check if there was an error while writing the script.
	  if ($?) {
	      $self->error("error writing data file $ddir/$fname: $!");
	      return 0;
	  }
      }
  }

  return 1;
}

=pod

=head2 Private methods

=item _initialize($comp_name, $config)

object initialization (done via new)

=cut

sub _initialize {
  my ($self,$name,$config)=@_;
  $self->setup_reporter();
  unless (defined $name && defined $config) {
    throw_error('bad initialization');
    return undef;
  }

  unless ($name =~ m{^([a-zA-Z_]\w+)$}) {
    throw_error ("Bad component name: $name");
    return undef;
  }

  $self->{'NAME'}=$1;
  $self->{'CONFIG'}=$config;

  # check for existing and 'active' in node profile

  # component itself does however not get loaded yet (this is done on demand
  # by the 'execute' commands)

  my $cdb_entry=$self->{'CONFIG'}->getElement('/software/components/'.$name);
  unless (defined $cdb_entry) {
    $ec->ignore_error();
    $self->error('no such component in node profile: '.$name);
    return undef;
  }

  my $prop=$config->getElement($_COMP_PREFIX.'/'.$name.'/active');
  unless (defined $prop) {
    $ec->ignore_error();
    $self->error('component '.$name.
		 " 'active' flag not found in node profile");
    return undef;
  } else {
    my $active=$prop->getBooleanValue();
    if ($active ne 'true') {
      $self->error('component '.$name.' is not active');
      return undef;
    }

    return ($self->_setDependencies());
  }
}


=pod

=item _load(): boolean

loads the component file in a separate namespace (NCM::Component::$name)

=cut

sub _load {
  my $self=shift;

  my $compname=$self->{'NAME'};

  # try to create the component from configuration information
  # or check that it is pre-installed
  if (!$self->writeComponent()) {
      unless ($self->hasFile()) {
	  $self->error('component '.$compname.' is not installed in /var/ncm/lib/perl/NCM/Component or /usr/lib/perl/NCM/Component');
	  return undef;
      }
  }

  eval ("use NCM::Component::$compname;");
  if ($@) {
    $self->error("bad Perl code in NCM::Component::$compname : $@");
    return undef;
  }

  my $comp_EC;
  eval "\$comp_EC=\$NCM::Component::$compname\:\:EC;";
  if ($@ || !defined $comp_EC || ref($comp_EC) ne 'LC::Exception::Context') {
    $self->error('bad component exception handler: $EC is not defined, not accessible or not of type LC::Exception::Context');
    $self->error("(note 1: the component package name has to be exactly 'NCM::Component::$compname' - please verify this inside '/usr/lib/perl/NCM/Component/$compname.pm' or '/var/ncm/lib/perl/NCM/Component/$compname.pm')");
    $self->error('(note 2: $EC has to be declared in "use vars (...)")');
    return undef;
  }

  my $component;
  eval("\$component=NCM::Component::$compname->new(\$compname, \$self)");
  if ($@) {
    $self->error("component $compname instantiation statement fails: $@");
    return undef;
  }
  return $component;
}


=pod

=item _setDependencies(): boolean

Reads the dependencies on other components via the NVA API and stores
them internally. They can be recovered by getDependencies()

=cut

sub _setDependencies {

  my ($self)=@_;

  $self->{'PRE_DEPS'}=[()];
  $self->{'POST_DEPS'}=[()];

  my $conf=$self->{'CONFIG'};

  my $pre_path = $_COMP_PREFIX.'/'.$self->{'NAME'}.'/dependencies/pre';
  my $post_path = $_COMP_PREFIX.'/'.$self->{'NAME'}.'/dependencies/post';


  # check if paths are defined (otherwise, no dependencies)


  my $res=$conf->getElement($pre_path);
  if (defined $res) {
    my $el;
    foreach $el ($res->getList()) {
      push (@{$self->{'PRE_DEPS'}},$el->getStringValue());
    }
    $self->debug(2,'pre dependencies for component '.$self->{'NAME'}.
		 ' '.join(',',@{$self->{'PRE_DEPS'}}));
  } else {
    $ec->ignore_error();
    $self->debug(1,'no pre dependencies found for '.$self->{'NAME'});
  }

  $res=$conf->getElement($post_path);
  if (defined $res) {
    my $el;
    foreach $el ($res->getList()) {
      push (@{$self->{'POST_DEPS'}},$el->getStringValue());
    }
    $self->debug(2,'post dependencies for component '.$self->{'NAME'}.
		 ' '.join(',',@{$self->{'POST_DEPS'}}));
  } else {
    $ec->ignore_error();
    $self->debug(1,'no post dependencies found for '.$self->{'NAME'});
  }


  return SUCCESS;
}



=pod

=item _execute

common function for executeConfigure() and executeUnconfigure()

=cut

sub _execute {
  my ($self,$method)=@_;

  # load the component

  my $retval;
  my $name=$self->name();
  my $component=$self->_load();
  unless (defined $component) {
    $self->error('cannot load component: '.$name);
    return undef;
  }

  # redirect log file to component's log file
  if ($this_app->option('multilog')) {
    my $logfile=$this_app->option("logdir").'/component-'.$name.'.log';
    my $objlog=CAF::Log->new($logfile,'at');
    unless (defined $objlog) {
      $self->error('cannot open component log file: '.$logfile);
      return undef;
    }
    $self->set_report_logfile($objlog);
  } else {
    $self->set_report_logfile ($this_app->{LOG});
  }

  $self->log('-----------------------------------------------------------');

  my $lcNoAct=$LC::Check::NoAction;
  if ($this_app->option('noaction')) {
    $LC::Check::NoAction=1;
    my $compname=$self->{'NAME'};
    my $noact_supported=undef;
    eval "\$noact_supported=\$NCM::Component::$compname\:\:NoActionSupported;";
    if ($@ || !defined $noact_supported || !$noact_supported) {
      # noaction is not supported by the component, skip execution in fake mod
      $self->info("component $compname has NoActionSupported not defined or to false, skipping noaction run");
      $retval= {
		'WARNINGS'=>0,
		'ERRORS'=>0
	       };
      return $retval;
    } else {
      $self->info("note: running component $compname in noaction mode");
    }
  }


  # execute component
  my $result;
  chdir ('/tmp');
  NCM::Template->SetDelimiters(@_TEMPLATE_DELIMITERS);
  my %ENV_BK=%ENV;
  # ensure that these signals get ignored
  # (%SIG is redefined by Perl itself sometimes)
  $SIG{$_}='IGNORE' foreach qw(HUP PIPE ALRM);

  # go and run the component
  eval "\$result=\$component->$method(\$self->{'CONFIG'});";

  %ENV=%ENV_BK; # restore env and signals
  if ($@) {
    $self->error("component ".$name.
		" executing method $method fails: $@");
    $retval=undef;
  } else {

    my $comp_EC;
    eval "\$comp_EC=\$NCM::Component::$name\:\:EC;";
    unless ($@) {
      if ($comp_EC->error) {
	$self->error('uncaught error exception in component:');
	my $formatter=$this_app->option('verbose') ||
	  $this_app->option('debug') ? "format_long" : "format_short";
	$component->error($comp_EC->error->$formatter());
	$comp_EC->ignore_error();
      }
      if ($comp_EC->warnings) {
	$self->warn('uncaught warning exception in component:');
	my $formatter=$this_app->option('verbose') ||
	  $this_app->option('debug') ? "format_long" : "format_short";
	foreach ($comp_EC->warnings) {
	  $component->warn($_->$formatter());
	}
	$comp_EC->ignore_warnings();
      }
    }

    if ($ec->error) {
      $self->error('error exception thrown by component:');
      my $formatter=$this_app->option('verbose') ||
	$this_app->option('debug') ? "format_long" : "format_short";
      $component->error($ec->error->$formatter());
      $ec->ignore_error();
    }
    if ($ec->warnings) {
      $self->warn('warning exception thrown by component:');
      my $formatter=$this_app->option('verbose') ||
	$this_app->option('debug') ? "format_long" : "format_short";
      foreach ($ec->warnings) {
	$component->warn($_->$formatter());
      }
      $ec->ignore_warnings();
    }

    # future: support a 'fatal' or 'abort' function
    #  if ($component->get_abort()) {
    #    $self->error("fatal error in component execution, aborting...");
    #    return undef;
    #  }

    $self->info('configure on component '.$component->name().' executed, '.
		$component->get_errors(). ' errors, '.
		$component->get_warnings(). ' warnings');
    $retval= {
	      'WARNINGS'=>$component->get_warnings(),
	      'ERRORS'=>$component->get_errors()
	     };
  }
  # restore logfile and noaction flags
  $self->set_report_logfile($this_app->{'LOG'})
    if ($this_app->option('multilog'));
  $LC::Check::NoAction=$lcNoAct;
  return $retval;
}





#+#############################################################################
1;

